examples/rust/battleship/src/main.rs

```diff
@@ -12,6 +12,210 @@
+    let mut number = String::new();
+    stdin().read_line(&mut number).expect("Failed to read line");
+    let count_players: u32 = number.trim().parse().expect("Not a number");
+    if count_players > 2 {
+        println!("Too many players...");
+    } else if count_players == 0 {
+        run_auto_bot();
+    } else {
+        let mut player1 = setup_player(1);
+        let mut player2 = setup_player(2);
+        let player1_init = player1.battleship.init().unwrap();
+        let player2_init = player2.battleship.init().unwrap();
+        player1.battleship.on_init_msg(&player2_init).unwrap();
+        player2.battleship.on_init_msg(&player1_init).unwrap();
+
+        for n in 1..10 {
+            if n % 2 == 0 {
+                println!(
+                    "{} turn, what are your attack coordiantes? (i.e.: 0,0)",
+                    player1.name
+                );
+                let mut attack_coord1 = String::new();
+                stdin()
+                    .read_line(&mut attack_coord1)
+                    .expect("Failed to read line");
+                let vec: Vec<&str> = attack_coord1.split(",").collect();
+
+                let j = round(
+                    &mut player1.battleship,
+                    &mut player2.battleship,
+                    vec[0].trim().parse::<u32>().unwrap(),
+                    vec[1].trim().parse::<u32>().unwrap(),
+                );
+                println!("It's a {:?}", j);
+            } else {
+                println!(
+                    "{} turn, what are your attack coordiantes? (i.e.: 0,0)",
+                    player2.name
+                );
+                let mut attack_coord2 = String::new();
+                stdin()
+                    .read_line(&mut attack_coord2)
+                    .expect("Failed to read line");
+                let vec: Vec<&str> = attack_coord2.split(",").collect();
+
+                let k = round(
+                    &mut player2.battleship,
+                    &mut player1.battleship,
+                    vec[0].trim().parse::<u32>().unwrap(),
+                    vec[1].trim().parse::<u32>().unwrap(),
+                );
+                println!("It's a {:?}", k);
+            }
+        }
+    }
+}
+
+fn setup_player(player: i32) -> Players {
+    println!("What's player {} name?", player);
+    let mut player_string = String::new();
+    stdin()
+        .read_line(&mut player_string)
+        .expect("Failed to read line");
+
+    let player_state = GameState {
+        ships: [
+            setup_ship("Carrier".to_string()),
+            setup_ship("Battleship".to_string()),
+            setup_ship("Cruiser".to_string()),
+            setup_ship("Submarine".to_string()),
+            setup_ship("Destroyer".to_string()),
+        ],
+        salt: 0xCAFECAFE,
+    };
+
+    return Players {
+        name: player_string,
+        battleship: Battleship::new(player_state),
+    };
+}
+
+fn setup_ship(ship_name: String) -> Ship {
```

Comment:
Does this need to be a `String`, I'm thinking this could be a `&str` (since we only need to borrow a reference to the string?).

Reply:
definitely need to stop thinking in terms of managed code 😅